scene/2d/animated_sprite_2d.cpp

```diff
@@ -160,79 +160,72 @@ void AnimatedSprite2D::_notification(int p_what) {
 			if (!frames->has_animation(animation)) {
 				return;
 			}
-			if (frame < 0) {
+			if (frames->get_animation_total_duration(animation) <= 0) {
 				return;
 			}
 
 			double remaining = get_process_delta_time();
-
+			int i = 0;
 			while (remaining) {
-				double speed = frames->get_animation_speed(animation) * speed_scale;
-				if (speed == 0) {
+				double actual_speed_scale = frames->get_animation_speed_scale(animation) * speed_scale;
+				if (backwards) {
+					actual_speed_scale *= -1;
+				}
+
+				if (actual_speed_scale == 0) {
 					return; // Do nothing.
 				}
 
+				int fc = frames->get_frame_count(animation);
+
 				if (timeout <= 0) {
-					timeout = _get_frame_duration();
+					// Next frame.
+					if (actual_speed_scale < 0) {
+						frame--;
+					} else {
+						frame++;
+					}
 
-					int last_frame = frames->get_frame_count(animation) - 1;
-					if (!backwards) {
-						// Forward.
-						if (frame >= last_frame) {
-							if (frames->get_animation_loop(animation)) {
-								frame = 0;
-								emit_signal(SceneStringNames::get_singleton()->animation_finished);
-							} else {
-								frame = last_frame;
-								if (!is_over) {
-									is_over = true;
-									emit_signal(SceneStringNames::get_singleton()->animation_finished);
-								}
-							}
+					if (frame < 0 || frame >= fc) {
+						if (frames->get_animation_loop(animation)) {
+							frame = (frame >= fc) ? 0 : fc - 1; // Wrap.
+							emit_signal(SceneStringNames::get_singleton()->animation_finished);
 						} else {
-							frame++;
-						}
-					} else {
-						// Reversed.
-						if (frame <= 0) {
-							if (frames->get_animation_loop(animation)) {
-								frame = last_frame;
+							frame = (frame >= fc) ? fc - 1 : 0; // Stop.
+							if (!is_over) {
+								is_over = true;
 								emit_signal(SceneStringNames::get_singleton()->animation_finished);
-							} else {
-								frame = 0;
-								if (!is_over) {
-									is_over = true;
-									emit_signal(SceneStringNames::get_singleton()->animation_finished);
-								}
 							}
-						} else {
-							frame--;
 						}
 					}
 
+					timeout = _get_frame_duration();
+
 					queue_redraw();
 
 					emit_signal(SceneStringNames::get_singleton()->frame_changed);
 				}
 
-				double to_process = MIN(timeout, remaining);
+				double to_process = MIN(timeout / Math::abs(actual_speed_scale), remaining);
+				timeout -= to_process * Math::abs(actual_speed_scale);
 				remaining -= to_process;
-				timeout -= to_process;
+
+				i++;
+				if (i > fc) {
+					return;
+				}
```

Comment:
Don't misunderstand. Just as I said in that comment, I'm just suggesting that the current branching of this PR is not a good way and that you change it to use the same branching method to advance the frame dependent on the duration.

Reply:
I have reasons for changing this code:

1\. `_get_frame_duration()` moved below because it returns the duration of the current frame. It was the same as long as the frames were the same вгкфешщт, but now without this change there is a bug that the вгкфешщт of the previous frame is used.

2\. A limit on the number of iterations is needed to prevent freezes if, for some reason, `to_process` is each time much less than `remaining`. A similar protection is implemented here:

https://github.com/godotengine/godot/blob/44c55ad59b01dbf360777b1a9a18a229aeb1f4f7/scene/resources/texture.cpp#L2622-L2623

3\. In my opinion, my code looks more compact and clear (lines 178-211 -> 182-200).

4\. And a couple more points.

I respect your contribution to Godot and your experience (of course, much more than mine), but understand me: I spent more than 12 hours on this PR, so you can not reject any part without understanding why something was done the way it was. If you have more precise comments and suggestions for improving this code, I will be glad.

Reply:
Hey, I see no reason to assume that Tokage just hastily rejected your changes without any consideration or understanding. But thank you for elaborating why you feel this is a required change.

Reply:
I don't object to 1 and 2 of your explanation, but I don't see any evidence for 3 and 4. I know all too rebasing is tired, but changes should be kept to a minimum.

Reply:
Also, as I mentioned in https://github.com/godotengine/godot/pull/65148#discussion_r968599167, I think it would be simpler to implement an internal backward flag than to implement the `actual_speed_scale`.

Reply:
3\. Compare:

<details>
<summary>Details</summary>

```cpp

int last_frame = frames->get_frame_count(animation) - 1;
if (!backwards) {
    // Forward.
    if (frame >= last_frame) {
        if (frames->get_animation_loop(animation)) {
            frame = 0;
            emit_signal(SceneStringNames::get_singleton()->animation_finished);
        } else {
            frame = last_frame;
            if (!is_over) {
                is_over = true;
                emit_signal(SceneStringNames::get_singleton()->animation_finished);
            }
        }
    } else {
        frame++;
    }
} else {
    // Reversed.
    if (frame <= 0) {
        if (frames->get_animation_loop(animation)) {
            frame = last_frame;
            emit_signal(SceneStringNames::get_singleton()->animation_finished);
        } else {
            frame = 0;
            if (!is_over) {
                is_over = true;
                emit_signal(SceneStringNames::get_singleton()->animation_finished);
            }
        }
    } else {
        frame--;
    }
}
```

and

```cpp


// Next frame.
if (actual_speed_scale < 0) {
    frame--;
} else {
    frame++;
}

if (frame < 0 || frame >= fc) {
    if (frames->get_animation_loop(animation)) {
        frame = (frame >= fc) ? 0 : fc - 1; // Wrap.
        emit_signal(SceneStringNames::get_singleton()->animation_finished);
    } else {
        frame = (frame >= fc) ? fc - 1 : 0; // Stop.
        if (!is_over) {
            is_over = true;
            emit_signal(SceneStringNames::get_singleton()->animation_finished);
        }
    }
}
```

</details>

Everyone has different tastes, but it seems to me that the code without duplication here looks better.

4\. This is the "and so on" item. I can write what I have not listed. But not now, sorry.

> Also, as I mentioned in [#65148 (comment)](https://github.com/godotengine/godot/pull/65148#discussion_r968599167), I think it would be simpler to implement an internal backward flag than to implement the `actual_speed_scale`.

It seems to me that this parameter should be removed as negative `speed_scale` is now supported.

@YuriSizov

<details>
<summary>Meta offtopic</summary>

I'm not making any assumptions about TokageItLab's motives (but I don't think they meant to offend me). However, it was definitely annoying to see a large selection of your work and under it a short comment "this should not be changed". Should not be changed because..?

I believe that such language should not be used in a review. If you do not agree, write why you think that this is not necessary or wrong. If you do not understand why this is needed, ask the PR author to explain why they think it is necessary or correct.

</details>

Reply:
> However, it was definitely annoying to see a large selection of your work and under it a short comment "this should not be changed". Should not be changed because..?
> 
> I believe that such language should not be used in a review. If you do not agree, write why you think that this is not necessary or wrong. If you do not understand why this is needed, ask the PR author to explain why they think it is necessary or correct.

Tokage's comment was clear about why it shouldn't be changed: because it was already fixed in another recent PR and your changes in that part are no longer necessary. I have no opinion on the code here to offer my technical point of view, but I don't see why you would think this is not enough to at least start the discussion.

Reply:
Please be aware that shorter code is not always better code. In the code you have shown, there is a redundant branch called `frame >= fc`, which is not so readable and actually increases the computational cost.

Reply:
Done.